scanner/gobin/gobin.go

```diff
@@ -0,0 +1,326 @@
+			IMAGE_SCN_MEM_WRITE            = 0x80000000
+			IMAGE_SCN_ALIGN_32BYTES        = 0x600000
+		)
+		for _, sect := range e.pe.Sections {
+			if sect.VirtualAddress != 0 && sect.Size != 0 &&
+				sect.Characteristics&^IMAGE_SCN_ALIGN_32BYTES == IMAGE_SCN_CNT_INITIALIZED_DATA|IMAGE_SCN_MEM_READ|IMAGE_SCN_MEM_WRITE {
+				return uint64(sect.VirtualAddress) + e.peBase
+			}
+		}
+	}
+	return 0
+}
+
+func (e *exe) ReadAddr(addr uint64, b []byte) (int, error) {
+	switch {
+	case e.elf != nil:
+		for _, p := range e.elf.Progs {
+			if p.Vaddr <= addr && addr <= p.Vaddr+p.Filesz-1 {
+				return p.ReadAt(b, int64(addr-p.Vaddr))
+			}
+		}
+	case e.pe != nil:
+		addr -= e.peBase
+		for _, s := range e.pe.Sections {
+			if uint64(s.VirtualAddress) <= addr && addr <= uint64(s.VirtualAddress+s.Size-1) {
+				return s.ReadAt(b, int64(addr-uint64(s.VirtualAddress)))
+			}
+		}
+	default:
+		return -1, errors.New("unsupported executable format")
+	}
+	return -1, errors.New("address not mapped")
+}
+
+func (e *exe) DecodeAs(blob []byte) error {
+	e.ptrSz = int(blob[14])
+	e.bin = binary.LittleEndian
+	if blob[15] != 0 {
+		e.bin = binary.BigEndian
+	}
+	e.ptr = e.bin.Uint64
+	if e.ptrSz == 4 {
+		e.ptr = func(b []byte) uint64 {
+			return uint64(e.bin.Uint32(b))
+		}
+	}
+	return nil
+}
+
+func (e *exe) StringAt(ea []byte) (string, error) {
+	h := make([]byte, 2*e.ptrSz)
+	n, err := e.ReadAddr(e.ptr(ea), h)
+	if err != nil {
+		return "", err
+	}
+	if n < len(h) {
+		return "", errors.New("short read")
+	}
+
+	addr, sz := e.ptr(h), e.ptr(h[e.ptrSz:])
+
+	out := make([]byte, sz)
+	n, err = e.ReadAddr(addr, out)
+	if err != nil {
+		return "", err
+	}
+	if n < len(out) {
+		return "", errors.New("short read")
+	}
+
+	return string(out), nil
+}
+
+const toobig = 30 * (1024 * 1024)       // 30MiB
+var magic = []byte("\xff Go buildinf:") // go magic build header
+
+// Scan performs a package scan on the given layer and returns all
+// the found packages
+func (s *Scanner) Scan(ctx context.Context, l *claircore.Layer) ([]*claircore.Package, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer", l.Hash.String())
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "scanner/gobin/Scanner.Scan"),
+		label.String("version", s.Version()),
+		label.Stringer("layer", l.Hash))
+	zlog.Debug(ctx).Msg("start")
+	defer zlog.Debug(ctx).Msg("done")
+
+	r, err := l.Reader()
+	if err != nil {
+		return nil, err
+	}
+	rd, ok := r.(interface {
+		io.ReadCloser
+		io.Seeker
+	})
+	if !ok {
+		return nil, errors.New("gobin: cannot seek on returned layer Reader")
+	}
+
+	var exes []exe
+	tr := tar.NewReader(rd)
+	var h *tar.Header
+	for h, err = tr.Next(); err == nil; h, err = tr.Next() {
+		fi := h.FileInfo()
+		m := fi.Mode()
+		switch {
+		case !m.IsRegular():
+			continue
+		case m.Perm()&0555 == 0:
+			// Not executable
+			continue
+		case fi.Size() > toobig:
```

Comment:
Curious if an arbitrary file size limit is necessary here? 

We don't care about reading the entire binary file, only the ELF header. With some finagling of the ELF header fields I think you can get the size of the elf header: 

```
      e_phnum
              This member holds the number of entries in the program
              header table.  Thus the product of e_phentsize and e_phnum
              gives the table's size in bytes.  If a file has no program
              header, e_phnum holds the value zero.
```

I'm thinking at that point, you don't care about the full Golang binary, you just slice off the ELF header bits.

Reply:
Could also reduce the likelihood of that by peaking for the elf/pe header

Reply:
That's along the lines I'm getting at here, if you can deduce the header sizes, you should be able to peak, and then only read out the portion of the file you need. I had the assumption this data is kept in something like PT_NOTE, but not too sure. I imagine the headers give you some information on where to peak/truncate/offset into to find what you need. 

I don't think this arbitrary limit is an issue, but if you can nix it somehow, it would be nice.